sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala

```diff
@@ -156,36 +165,90 @@ case class Window(
-      new UnboundedWindowFunctionFrame(ordinal, functions)
-
-    // Error
-    case fr =>
-      sys.error(s"Unsupported Frame $fr for functions: $functions")
+    frame: (Char, WindowFrame),
+    functions: Array[Expression],
+    ordinal: Int,
+    target: MutableRow): WindowFunctionFrame = {
+
+    // Construct an aggregate processor if we have to.
+    def processor = {
+      val prepared = functions.map(_.asInstanceOf[AggregateFunction])
+      AggregateProcessor(prepared, ordinal, child.output, newMutableProjection)
+    }
+
+    // Create the frame processor.
+    frame match {
+      // Offset Frame
+      case ('O', SpecifiedWindowFrame(RowFrame,
+                  FrameBoundaryExtractor(l),
+                  FrameBoundaryExtractor(h)))
+        if l == h =>
+        new OffsetWindowFunctionFrame(target, ordinal, functions, child.output,
+          newMutableProjection, l)
+
+      // Growing Frame.
+      case ('A', SpecifiedWindowFrame(frameType,
+                  UnboundedPreceding,
+                  FrameBoundaryExtractor(high))) =>
+        val uBoundOrdering = createBoundOrdering(frameType, high)
+        new UnboundedPrecedingWindowFunctionFrame(target, processor, uBoundOrdering)
+
+      // Shrinking Frame.
+      case ('A', SpecifiedWindowFrame(frameType,
+                  FrameBoundaryExtractor(low),
+                  UnboundedFollowing)) =>
+        val lBoundOrdering = createBoundOrdering(frameType, low)
+        new UnboundedFollowingWindowFunctionFrame(target, processor, lBoundOrdering)
+
+      // Moving Frame.
+      case ('A', SpecifiedWindowFrame(frameType,
+                  FrameBoundaryExtractor(l),
+                  FrameBoundaryExtractor(h))) =>
+        val lBoundOrdering = createBoundOrdering(frameType, l)
+        val uBoundOrdering = createBoundOrdering(frameType, h)
+        new SlidingWindowFunctionFrame(target, processor, lBoundOrdering, uBoundOrdering)
+
+      // Entire Partition Frame.
+      case ('A', SpecifiedWindowFrame(_,
+                  UnboundedPreceding,
+                  UnboundedFollowing)) =>
+        new UnboundedWindowFunctionFrame(target, processor)
+    }
+  }
+
+  /** Map of window expressions and functions by their key. */
+  private[this] lazy val windowFunctionMap: FunctionMap = {
+    val functions: FunctionMap = mutable.Map.empty
+
+    // Add a function and its function to the map for a given frame.
+    def collect(tpe: Char, frame: WindowFrame, e: Expression, f: Expression): Unit = {
+      val (es, fs) = functions.getOrElseUpdate(
+        (tpe, frame), (ArrayBuffer.empty[Expression], ArrayBuffer.empty[Expression]))
+      es.append(e)
+      fs.append(f)
+    }
+
+    // Collect all valid window functions.
+    windowExpression.foreach { x =>
+      x.foreach {
+        case e @ WindowExpression(function, spec) =>
+          val frame = spec.frameSpecification
+          function match {
+            case AggregateExpression(f, _, _) => collect('A', frame, e, f)
```

Comment:
It'll help. I am not really happy with this entire code block; further suggestions are welcome.